src/shared/dlfcn-util.c

```diff
@@ -2,39 +2,63 @@
 
 #include "dlfcn-util.h"
 
-int dlsym_many_and_warn(void *dl, int level, ...) {
-        va_list ap;
-        int r;
-
-        /* Tries to resolve a bunch of function symbols, and logs errors about the ones it cannot
-         * resolve. Note that this function possibly modifies the supplied function pointers if the whole
-         * operation fails */
+static int dlsym_many_or_warnv(void *dl, int log_level, va_list ap) {
+        void (**fn)(void);
 
-        va_start(ap, level);
+        /* Tries to resolve a bunch of function symbols, and logs an error about if it cannot resolve one of
+         * them. Note that this function possibly modifies the supplied function pointers if the whole
+         * operation fails. */
 
-        for (;;) {
-                void (**fn)(void);
+        while ((fn = va_arg(ap, typeof(fn)))) {
                 void (*tfn)(void);
                 const char *symbol;
 
-                fn = va_arg(ap, typeof(fn));
-                if (!fn)
-                        break;
-
                 symbol = va_arg(ap, typeof(symbol));
 
                 tfn = (typeof(tfn)) dlsym(dl, symbol);
-                if (!tfn) {
-                        r = log_full_errno(level,
-                                           SYNTHETIC_ERRNO(ELIBBAD),
-                                           "Can't find symbol %s: %s", symbol, dlerror());
-                        va_end(ap);
-                        return r;
-                }
-
+                if (!tfn)
+                        return log_full_errno(log_level,
+                                              SYNTHETIC_ERRNO(ELIBBAD),
+                                              "Can't find symbol %s: %s", symbol, dlerror());
                 *fn = tfn;
         }
 
-        va_end(ap);
         return 0;
 }
+
+int dlsym_many_or_warn(void *dl, int log_level, ...) {
+        va_list ap;
+        int r;
+
+        va_start(ap, log_level);
+        r = dlsym_many_or_warnv(dl, log_level, ap);
+        va_end(ap);
+
+        return r;
+}
+
+int dlopen_many_sym_or_warn_sentinel(void **ret_dl, const char *filename, int log_level, ...) {
```

Comment:
`ret_` is supposed to be the prefix for return variables, i.e. output params. but this one is both input *and* output. maybe name dlp?

Reply:
Makes sense.